lightningd/onchain_control.c

```diff
@@ -841,11 +845,172 @@ static bool consider_onchain_rebroadcast(struct channel *channel,
 	return true;
 }
 
+static bool meets_feerate(const struct bitcoin_tx *tx, u32 feerate)
+{
+	return tx_feerate(tx) >= feerate;
+}
+
+static int find_change_output(const struct bitcoin_tx *tx,
+			      const struct onchain_signing_info *info)
+{
+	for (size_t i = 0; i < tx->wtx->num_outputs; i++) {
+		if (elements_tx_output_is_fee(tx, i))
+			continue;
+		if (i == info->htlc_inout)
+			continue;
+		return i;
+	}
+	return -1;
+}
+
 static bool consider_onchain_htlc_tx_rebroadcast(struct channel *channel,
 						 const struct bitcoin_tx **tx,
 						 struct onchain_signing_info *info)
 {
-	/* FIXME: Implement rbf! */
+	struct amount_sat fee, excess, change;
+	u32 feerate;
+	struct bitcoin_tx *newtx;
+	int change_output;
+	size_t locktime;
+	struct bitcoin_txid oldtxid, newtxid;
+	struct wally_psbt *psbt;
+	struct utxo **utxos;
+	const u8 *msg;
+	struct lightningd *ld = channel->peer->ld;
+
+	/* We can't do much without anchor outputs (we could CPFP?) */
+	if (!channel_has(channel, OPT_ANCHOR_OUTPUTS))
+		return true;
+
+	/* Already paying enough? */
+	feerate = feerate_for_target(ld->topology, info->deadline_block);
+	if (meets_feerate(*tx, feerate))
+		return true;
+
+	/* Make a copy to play with */
+	newtx = clone_bitcoin_tx(tmpctx, *tx);
+
+	/* If we have a change output already, try reducing that. */
+	change_output = find_change_output(newtx, info);
+	if (change_output != -1) {
+		struct amount_sat amt, target_fee, needed;
+
+		target_fee = amount_sat(feerate * (u64)bitcoin_tx_weight(newtx) / 1000);
+		/* Shouldn't happen, that we already exceed fees needed. */
+		if (!amount_sat_sub(&needed,
+				    target_fee, bitcoin_tx_compute_fee(newtx)))
+			return true;
+		bitcoin_tx_output_get_amount_sat(newtx, change_output, &amt);
+		if (!amount_sat_sub(&amt, amt, needed)) {
+			/* Eliminate this output altogether. */
+			bitcoin_tx_remove_output(newtx, change_output);
+		} else {
+			bitcoin_tx_output_set_amount(newtx, change_output, amt);
+		}
+	}
+
+	if (meets_feerate(newtx, feerate)) {
+		psbt = newtx->psbt;
+		goto replace;
+	}
+
+	/* OK, we need to attach another input! */
+	/* FIXME: Maybe more than one! */
+	utxos = tal_arr(tmpctx, struct utxo *, 1);
+	utxos[0] = wallet_find_utxo(utxos, ld->wallet,
```

Comment:
If so, this may allow an attacker to dust you wallet to slow down progress at picking a "good enough" utxo. Might want to just keep grabbing utxos rather than aborting at a min?